### PR TITLE
110600 - CompletionContext MSI retrieval

### PR DESCRIPTION
**Pull request by agustarnason**
Connected to [AB#110600](https://statoildeveloper.visualstudio.com/b6d97460-2e53-47ae-aad4-b41396d90828/_workitems/edit/110600)

Updated to fetching CompletionContext from MSI instead of having that information saved locally

* Added MSI Access Token Provider
* Updated CompletionContext to MSI Access Token and Token Credentials when creating its connection
* Ran the following commands on the dev database to allow access via MSI:
```
CREATE USER [App Name] FROM EXTERNAL PROVIDER;

ALTER ROLE db_datareader ADD MEMBER [App Name];
ALTER ROLE db_datawriter ADD MEMBER [App Name];
```

### CODE REVIEW

**Comment by janid1967** [comment]
Add description of manual steps into #110523

File: src/Equinor.ProCoSys.Completion.Infrastructure/MsiAccessTokenProvider.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace Equinor.ProCoSys.Completion.Infrastructure;
+public static class MsiAccessTokenProvider
+{
+    public static AccessToken? AccessToken { get; private set; }
+
+    public static async Task<string> GetAccessTokenAsync(TokenCredential credential)
+    {
+        if (AccessToken.HasValue && AccessToken.Value.ExpiresOn > DateTimeOffset.UtcNow)
+            return AccessToken.Value.Token;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Infrastructure/MsiAccessTokenProvider.cs:14]: Apply braces style
image
(check other files for same issue)

File: src/Equinor.ProCoSys.Completion.Infrastructure/CompletionContext.cs [hunk 2]
```diff
@@ -39,12 +41,20 @@ public class CompletionContext : DbContext, IUnitOfWork, IReadOnlyContext
...
         : base(options)
     {
         _plantProvider = plantProvider;
         _eventDispatcher = eventDispatcher;
         _currentUserProvider = currentUserProvider;
+
+        if (Database?.ProviderName != "Microsoft.EntityFrameworkCore.SqlServer")
+            return;
+
+        var connection = Database.GetDbConnection() as SqlConnection;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Infrastructure/CompletionContext.cs:55]: @agustarnason Can we discuss this? Both code-style and the logic itself?

**Comment by agustarnason** [comment]
> Add description of manual steps into #110523

Added
Let me know if what I wrote is unclear or if there is specific information you think I forgot or left out

File: src/Equinor.ProCoSys.Completion.Infrastructure/MsiAccessTokenProvider.cs (hunk 1 above)
Comment (agustarnason) [reply src/Equinor.ProCoSys.Completion.Infrastructure/MsiAccessTokenProvider.cs:15]: Updated this function to add the curly braces

File: src/Equinor.ProCoSys.Completion.Infrastructure/CompletionContext.cs (hunk 2 above)
Comment (agustarnason) [reply src/Equinor.ProCoSys.Completion.Infrastructure/CompletionContext.cs]: Replaced the code with:

        // Should not run during integration tests
        if (Database is object
            && Database.ProviderName == "Microsoft.EntityFrameworkCore.SqlServer"
            && Database.GetDbConnection() is SqlConnection connection)
        {
            connection.AccessToken = MsiAccessTokenProvider.GetAccessTokenAsync(credential).Result;
        }
        
 Let me know if you'd like other changes made here

Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.Infrastructure/CompletionContext.cs]: I've done it together with @DownGoat ... I resolve this conversation

File: src/Equinor.ProCoSys.Completion.WebApi/Startup.cs [hunk 3]
```diff
@@ -66,6 +68,18 @@ public void ConfigureServices(IServiceCollection services)
             }
         }
 
+        TokenCredential credential = _environment.IsDevelopment() switch
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Startup.cs:71]: Suggestion: Add a comment why we do this (@DownGoat explained it, but ok to have it written for other developers too ... your choice)

**Review by janid1967 — APPROVED** [review]

File: src/Equinor.ProCoSys.Completion.WebApi/Startup.cs (hunk 3 above)
Comment (agustarnason) [reply src/Equinor.ProCoSys.Completion.WebApi/Startup.cs:75]: Added a comment here explaining how credential selection works when run locally/in development

Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.WebApi/Startup.cs:75]: @agustarnason You don't have to wait for yet an approval after pushing minors as this. Just do it, push, resolve conversation and merge the PR. I don't need respons on comment either. Just do it ;-)

**Pull request merged**